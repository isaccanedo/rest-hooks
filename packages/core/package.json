{
  "name": "@rest-hooks/core",
  "version": "3.4.0",
  "description": "Asynchronous data framework for React",
  "sideEffects": false,
  "main": "dist/index.js",
  "react-native": "legacy/index.js",
  "module": "legacy/index.js",
  "unpkg": "dist/index.umd.min.js",
  "types": "lib/index.d.ts",
  "typesVersions": {
    ">=4.0": {
      "": [
        "lib/index.d.ts"
      ],
      "*": [
        "lib/index.d.ts"
      ]
    },
    ">=3.4": {
      "": [
        "ts3.4/index.d.ts"
      ],
      "*": [
        "ts3.4/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": {
      "module": "./lib/index.js",
      "import": "./node.mjs",
      "require": "./dist/index.js",
      "default": "./lib/index.js"
    },
    "./package.json": "./package.json"
  },
  "type": "module",
  "engines": {
    "node": "^12.17 || ^13.7 || >=14"
  },
  "files": [
    "src",
    "dist",
    "lib",
    "ts3.4",
    "node.mjs",
    "legacy",
    "LICENSE",
    "README.md",
    "./typescript.svg",
    "./rest_hooks_logo_and_text.svg"
  ],
  "scripts": {
    "build:lib": "NODE_ENV=production BROWSERSLIST_ENV='2020' babel --root-mode upward src --out-dir lib --source-maps inline --extensions '.ts,.tsx,.js,.jsx' --ignore '**/__tests__/**' --ignore '**/*.d.ts'",
    "build:legacy:lib": "NODE_ENV=production BROWSERSLIST_ENV='2018' babel --root-mode upward src --out-dir legacy --source-maps inline --extensions '.ts,.tsx,.js,.jsx' --ignore '**/__tests__/**' --ignore '**/*.d.ts'",
    "build:js:node": "BROWSERSLIST_ENV=node12 rollup -c",
    "build:js:browser": "BROWSERSLIST_ENV=legacy rollup -c",
    "build:bundle": "run-s build:js:\\* && echo '{\"type\":\"commonjs\"}' > dist/package.json",
    "build:clean": "rimraf lib ts3.4 legacy dist *.tsbuildinfo ./index.d.ts",
    "build:legacy-types": "yarn run downlevel-dts lib ts3.4 && copyfiles --up 1 ./src-legacy-types/**/*.d.ts ./ts3.4/",
    "build": "yarn run build:lib && yarn run build:legacy:lib && yarn run build:bundle",
    "dev": "yarn run build:lib -w",
    "prepare": "yarn run build:lib",
    "prepack": "yarn prepare",
    "prepublishOnly": "yarn run build:bundle && yarn run build:legacy:lib"
  },
  "keywords": [
    "react",
    "data",
    "cache",
    "flux",
    "suspense",
    "fetch",
    "hook",
    "networking",
    "async",
    "concurrent mode",
    "typescript",
    "async",
    "data fetching",
    "data cache",
    "api client",
    "api",
    "normalized cache",
    "swr",
    "GraphQL",
    "query",
    "react-native",
    "ios",
    "android",
    "web",
    "middleware",
    "declarative",
    "scalable"
  ],
  "author": "Nathaniel Tucker <me@ntucker.me> (https://github.com/ntucker)",
  "license": "Apache-2.0",
  "homepage": "https://resthooks.io",
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:coinbase/rest-hooks.git",
    "directory": "packages/core"
  },
  "bugs": {
    "url": "https://github.com/coinbase/rest-hooks/issues"
  },
  "dependencies": {
    "@babel/runtime": "^7.13.0",
    "@rest-hooks/normalizr": "^9.2.0",
    "@rest-hooks/use-enhanced-reducer": "^1.2.0",
    "flux-standard-action": "^2.1.1"
  },
  "peerDependencies": {
    "@types/react": "^16.8.4 || ^17.0.0 || ^18.0.0-0",
    "react": "^16.8.4 || ^17.0.0 || ^18.0.0-0"
  },
  "peerDependenciesMeta": {
    "@types/react": {
      "optional": true
    }
  },
  "devDependencies": {
    "@babel/cli": "7.19.3",
    "@babel/core": "7.20.2",
    "@types/babel__core": "^7",
    "downlevel-dts": "^0.10.0",
    "npm-run-all": "^4.1.5",
    "rollup": "2.79.1",
    "rollup-plugin-babel": "^4.4.0",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-dts": "^5.0.0",
    "rollup-plugin-filesize": "^9.1.2",
    "rollup-plugin-json": "^4.0.0",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-replace": "^2.2.0",
    "rollup-plugin-terser": "^7.0.2"
  }
}
