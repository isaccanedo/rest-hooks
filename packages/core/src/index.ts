export * as __INTERNAL__ from './internal.js';
export type {
  NetworkError,
  UnknownError,
  ErrorTypes,
  Schema,
  EndpointInterface,
  EntityInterface,
  ResolveType,
} from '@rest-hooks/normalizr';
export { ExpiryStatus } from '@rest-hooks/normalizr';
export {
  default as NetworkManager,
  ResetError,
} from './state/NetworkManager.js';
export {
  default as createReducer,
  initialState,
} from './state/createReducer.js';
export { default as reducer } from './state/reducerInstance.js';
export { default as applyManager } from './state/applyManager.js';
export { useDenormalized } from './state/selectors/index.js';
export {
  useCache,
  useFetcher,
  useController,
  useFetchDispatcher,
  useRetrieve,
  useResource,
  useSubscription,
  useMeta,
  useError,
  CacheProvider,
  BackupBoundary,
  useInvalidator,
  useInvalidateDispatcher,
  useResetter,
  hasUsableData,
} from './react-integration/index.js';
export {
  useSuspense,
  useFetch,
  useDLE,
} from './react-integration/newhooks/index.js';
export {
  StateContext,
  DispatchContext,
  DenormalizeCacheContext,
  ControllerContext,
  StoreContext,
  type Store,
} from './react-integration/context.js';
export { default as Controller } from './controller/Controller.js';

export * from './controller/types.js';
export * from './state/actions/index.js';
export * as actionTypes from './actionTypes.js';
export { usePromisifiedDispatch } from '@rest-hooks/use-enhanced-reducer';
/* istanbul ignore next */
export * from './types.js';
export type {
  FetchShape,
  ReadShape,
  MutateShape,
  DeleteShape,
} from './endpoint/shapes.js';
export type {
  SetShapeParams,
  ParamsFromShape,
  BodyFromShape,
  ReturnFromShape,
} from './endpoint/types.js';
