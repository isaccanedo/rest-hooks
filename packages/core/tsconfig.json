{
  "extends": "../../tsconfig-base",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src"
  },
  "include": ["src"],
  "references": [
    { "path": "../../__tests__" },
    { "path": "../normalizr" },
    { "path": "../use-enhanced-reducer" }
  ]
}
