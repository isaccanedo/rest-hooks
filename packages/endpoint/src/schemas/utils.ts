export const getValues = (input: any) =>
  Array.isArray(input) ? input : Object.keys(input).map(key => input[key]);

export const filterEmpty = ([item, , deletedItem]: any) =>
  item !== undefined && !deletedItem;
