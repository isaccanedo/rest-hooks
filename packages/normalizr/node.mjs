export * from './dist/normalizr.js';
