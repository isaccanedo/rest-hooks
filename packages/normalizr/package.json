{
  "name": "@rest-hooks/normalizr",
  "version": "9.2.0",
  "description": "Normalizes and denormalizes JSON according to schema for Redux and Flux applications",
  "homepage": "https://github.com/coinbase/rest-hooks/tree/master/packages/normalizr#readme",
  "bugs": {
    "url": "https://github.com/coinbase/rest-hooks/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:coinbase/rest-hooks.git",
    "directory": "packages/normalizr"
  },
  "keywords": [
    "flux",
    "redux",
    "normalize",
    "denormalize",
    "data",
    "cache",
    "api",
    "react-native",
    "ios",
    "android",
    "web",
    "json"
  ],
  "files": [
    "dist/",
    "lib/",
    "node.mjs",
    "legacy",
    "ts3.4",
    "LICENSE",
    "README.md"
  ],
  "main": "dist/normalizr.js",
  "react-native": "legacy/index.js",
  "module": "legacy/index.js",
  "types": "lib/index.d.ts",
  "typesVersions": {
    ">=4.0": {
      "": [
        "lib/index.d.ts"
      ],
      "*": [
        "lib/index.d.ts"
      ]
    },
    ">=3.4": {
      "": [
        "ts3.4/index.d.ts"
      ],
      "*": [
        "ts3.4/index.d.ts"
      ]
    }
  },
  "exports": {
    ".": {
      "module": "./lib/index.js",
      "import": "./node.mjs",
      "require": "./dist/normalizr.js",
      "default": "./lib/index.js"
    },
    "./package.json": "./package.json"
  },
  "type": "module",
  "engines": {
    "node": "^12.17 || ^13.7 || >=14"
  },
  "sideEffects": false,
  "scripts": {
    "build": "yarn run build:lib && yarn run build:legacy:lib && yarn run build:bundle",
    "build:js:development": "BROWSERSLIST_ENV=legacy NODE_ENV=development rollup -c",
    "build:js:production": "BROWSERSLIST_ENV=legacy NODE_ENV=production rollup -c",
    "build:js:node": "BROWSERSLIST_ENV=node12 NODE_ENV=development rollup -c",
    "build:lib": "NODE_ENV=production BROWSERSLIST_ENV='2020' babel --root-mode upward src --out-dir lib --source-maps inline --extensions '.ts,.tsx,.js' --ignore '**/__tests__/**' --ignore '**/*.d.ts'",
    "build:legacy:lib": "NODE_ENV=production BROWSERSLIST_ENV='2018' babel --root-mode upward src --out-dir legacy --source-maps inline --extensions '.ts,.tsx,.js' --ignore '**/__tests__/**' --ignore '**/*.d.ts'",
    "build:bundle": "run-s build:js:\\* && echo '{\"type\":\"commonjs\"}' > dist/package.json",
    "build:clean": "rimraf lib ts3.4 legacy dist *.tsbuildinfo",
    "build:legacy-types": "yarn run downlevel-dts lib ts3.4",
    "lint": "yarn lint:cmd --fix",
    "lint:ci": "yarn lint:cmd",
    "lint:cmd": "eslint . --ext '.js,.json,.snap' --cache",
    "precommit": "flow check && lint-staged",
    "prepublishOnly": "npm run build"
  },
  "author": "Nathaniel Tucker",
  "contributors": [
    "Paul Armstrong",
    "Dan Abramov"
  ],
  "license": "MIT",
  "devDependencies": {
    "@babel/cli": "7.19.3",
    "@babel/core": "7.20.2",
    "@types/babel__core": "^7",
    "downlevel-dts": "^0.10.0",
    "immutable": "4.1.0",
    "npm-run-all": "^4.1.5",
    "rollup": "2.79.1",
    "rollup-plugin-babel": "^4.4.0",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-filesize": "^9.1.2",
    "rollup-plugin-json": "^4.0.0",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-replace": "^2.2.0",
    "rollup-plugin-terser": "^7.0.2"
  },
  "dependencies": {
    "@babel/runtime": "^7.13.0",
    "npm-run-all": "^4.1.5"
  }
}
