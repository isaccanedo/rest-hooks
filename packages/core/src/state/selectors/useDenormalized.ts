import type { Schema, DenormalizeNullable } from '@rest-hooks/normalizr';
import { useMemo } from 'react';
import { ExpiryStatus } from '@rest-hooks/normalizr';

import { State } from '../../types.js';
import { ReadShape, ParamsFromShape } from '../../endpoint/index.js';
import useController from '../../react-integration/hooks/useController.js';
import shapeToEndpoint from '../../endpoint/adapter.js';

/**
 * @deprecated use https://resthooks.io/docs/api/Controller#getResponse directly instead
 *
 * Selects the denormalized form from `state` cache.
 *
 * If `result` is not found, will attempt to generate it naturally
 * using params and schema. This increases cache hit rate for many
 * detail shapes.
 *
 * @returns [denormalizedValue, ready]
 */
export default function useDenormalized<
  Shape extends Pick<
    ReadShape<Schema | undefined, any>,
    'getFetchKey' | 'schema' | 'options'
  >,
>(
  shape: Shape,
  params: ParamsFromShape<Shape> | null,
  state: State<any>,
  /** @deprecated */
  denormalizeCache?: any,
): {
  data: DenormalizeNullable<Shape['schema']>;
  expiryStatus: ExpiryStatus;
  expiresAt: number;
} {
  const controller = useController();

  const endpoint = useMemo(() => {
    return shapeToEndpoint(shape);
    // we currently don't support shape changes
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const key = params !== null ? endpoint.key(params) : '';
  const cacheResults = params && state.results[key];

  // Compute denormalized value
  return useMemo(() => {
    return controller.getResponse(endpoint, params, state) as {
      data: DenormalizeNullable<Shape['schema']>;
      expiryStatus: ExpiryStatus;
      expiresAt: number;
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    cacheResults,
    state.indexes,
    state.entities,
    state.entityMeta,
    key,
    cacheResults,
  ]);
}
