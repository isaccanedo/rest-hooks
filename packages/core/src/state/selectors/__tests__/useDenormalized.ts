import {
  CoolerArticleResource,
  PaginatedArticleResource,
  NestedArticleResource,
  UserResource,
  IndexedUserResource,
  photoShape,
} from '__tests__/legacy-3';
import { createEntityMeta } from '__tests__/utils';
import { normalize, NormalizedIndex } from '@rest-hooks/normalizr';
import { initialState } from '@rest-hooks/core';
import { renderHook, act } from '@testing-library/react-hooks';
import { useState } from 'react';
import { ExpiryStatus } from '@rest-hooks/normalizr';

import useDenormalized from '../useDenormalized';

describe('useDenormalized()', () => {
  describe('Single', () => {
    const params = { id: 5, title: 'bob', content: 'head' };
    const article = CoolerArticleResource.fromJS(params);
    describe('state is empty', () => {
      const state = initialState;
      const { result } = renderHook(() =>
        useDenormalized(CoolerArticleResource.detailShape(), { id: 5 }, state),
      );

      it('expiryStatus should be InvalidIfStale', () => {
        expect(result.current.expiryStatus).toBe(ExpiryStatus.InvalidIfStale);
      });

      it('should provide inferred results with undefined', () => {
        expect(result.current.data).toMatchInlineSnapshot(`undefined`);
      });
    });
    describe('state is populated just not with our query', () => {
      const state = {
        ...initialState,
        entities: {
          [CoolerArticleResource.key]: {
            [params.id]: article,
          },
        },
        results: {
          [CoolerArticleResource.detailShape().getFetchKey(params)]: params.id,
        },
      };
      const { result } = renderHook(() =>
        useDenormalized(
          CoolerArticleResource.detailShape(),
          {
            id: 543345345345453,
          },
          state,
        ),
      );

      it('expiryStatus should be InvalidIfStale', () => {
        expect(result.current.expiryStatus).toBe(ExpiryStatus.InvalidIfStale);
      });

      it('should provide inferred results with undefined', () => {
        expect(result.current.data).toMatchInlineSnapshot(`undefined`);
      });
    });
    describe('when state exists', () => {
      const state = {
        ...initialState,
        entities: {
          [CoolerArticleResource.key]: {
            [params.id]: article,
          },
        },
        results: {
          [CoolerArticleResource.detailShape().getFetchKey(params)]: params.id,
        },
      };
      state.entityMeta = createEntityMeta(state.entities);
      const {
        result: {
          current: { data, expiryStatus, expiresAt },
        },
      } = renderHook(() =>
        useDenormalized(CoolerArticleResource.detailShape(), params, state),
      );

      it('expiryStatus should be Valid', () => {
        expect(expiryStatus).toBe(ExpiryStatus.Valid);
      });

      it('should provide inferred results', () => {
        expect(data).toStrictEqual(article);
        expect(data).toBeInstanceOf(CoolerArticleResource);
      });
    });
    describe('without entity with defined results', () => {
      const state = {
        ...initialState,
        entities: { [CoolerArticleResource.key]: {} },
        results: {
          [CoolerArticleResource.detailShape().getFetchKey(params)]: params.id,
        },
      };
      const {
        result: {
          current: { data, expiryStatus, expiresAt },
        },
      } = renderHook(() =>
        useDenormalized(CoolerArticleResource.detailShape(), params, state),
      );

      it('expiryStatus should be Valid', () => {
        expect(expiryStatus).toBe(ExpiryStatus.Valid);
      });

      it('should provide inferred results with undefined', () => {
        expect(data).toMatchInlineSnapshot(`undefined`);
      });
    });
    describe('no result exists but primary key is used', () => {
      const state = {
        ...initialState,
        entities: {
          [CoolerArticleResource.key]: {
            [params.id]: article,
          },
        },
      };
      state.entityMeta = createEntityMeta(state.entities);
      const {
        result: {
          current: { data, expiryStatus, expiresAt },
        },
      } = renderHook(() =>
        useDenormalized(CoolerArticleResource.detailShape(), params, state),
      );

      it('expiryStatus should be Valid', () => {
        expect(expiryStatus).toBe(ExpiryStatus.Valid);
      });

      it('should provide inferred results', () => {
        expect(data).toStrictEqual(article);
        expect(data).toBeInstanceOf(CoolerArticleResource);
      });
    });
    describe('no result exists but primary key is used when using nested schema', () => {
      const pageArticle = PaginatedArticleResource.fromJS(article);
      const state = {
        ...initialState,
        entities: {
          [PaginatedArticleResource.key]: {
            [params.id]: pageArticle,
          },
        },
      };
      state.entityMeta = createEntityMeta(state.entities);
      const {
        result: {
          current: { data, expiryStatus, expiresAt },
        },
      } = renderHook(() =>
        useDenormalized(PaginatedArticleResource.detailShape(), params, state),
      );

      it('expiryStatus should be Valid', () => {
        expect(expiryStatus).toBe(ExpiryStatus.Valid);
      });

      it('should provide inferred results', () => {
        expect(data.data).toStrictEqual(pageArticle);
        expect(data.data).toBeInstanceOf(PaginatedArticleResource);
      });
    });

    describe('no result exists but index is used when using nested schema', () => {
      const pageArticle = PaginatedArticleResource.fromJS({
        ...params,
        author: IndexedUserResource.fromJS({ id: 23, username: 'anne' }),
      });
      const user = IndexedUserResource.fromJS({ id: 23, username: 'anne' });
      const IndexShape = {
        type: 'read' as const,
        getFetchKey({ username }: { username: string }) {
          return username;
        },
        schema: {
          pagination: { next: '', previous: '' },
          data: IndexedUserResource,
        },
      };

      it('should find value on index updates', () => {
        let localstate = {
          ...initialState,
          entities: {
            [PaginatedArticleResource.key]: {
              [`${pageArticle.pk()}`]: pageArticle,
            },
            [UserResource.key]: { [`${user.pk()}`]: user },
          },
        };
        localstate.entityMeta = createEntityMeta(localstate.entities);

        const { result, rerender } = renderHook(
          ({ state }) =>
            useDenormalized(IndexShape, { username: user.username }, state),
          { initialProps: { state: localstate } },
        );
        expect(result.current.expiryStatus).toBe(ExpiryStatus.InvalidIfStale);
        localstate = {
          ...localstate,
          indexes: {
            [IndexedUserResource.key]: {
              username: {
                [user.username]: user.pk(),
              },
            },
          } as NormalizedIndex,
        };
        rerender({ state: localstate });
        expect(result.current.expiryStatus).toBe(ExpiryStatus.Valid);
        expect(result.current.data.data).toStrictEqual(user);
      });
    });

    describe('not using primary key as param', () => {
      const urlParams = { title: 'bob' };
      const state = {
        ...initialState,
        entities: {
          [CoolerArticleResource.key]: {
            [params.id]: article,
          },
        },
        results: {
          [CoolerArticleResource.detailShape().getFetchKey(urlParams)]:
            params.id,
        },
      };
      state.entityMeta = createEntityMeta(state.entities);
      const {
        result: {
          current: { data, expiryStatus, expiresAt },
        },
      } = renderHook(() =>
        useDenormalized(CoolerArticleResource.detailShape(), params, state),
      );

      it('expiryStatus should be Valid', () => {
        expect(expiryStatus).toBe(ExpiryStatus.Valid);
      });

      it('should provide inferred results', () => {
        expect(data).toStrictEqual(article);
        expect(data).toBeInstanceOf(CoolerArticleResource);
      });
    });
    it('should throw when results are Array', () => {
      const params = { title: 'bob' };
      const state = {
        ...initialState,
        results: {
          [CoolerArticleResource.detailShape().getFetchKey(params)]: [5, 6, 7],
        },
      };
      const { result } = renderHook(() =>
        useDenormalized(CoolerArticleResource.detailShape(), params, state),
      );
      expect(result.error).toBeDefined();
    });
    it('should throw when results are Object', () => {
      const params = { title: 'bob' };
      const state = {
        ...initialState,
        results: {
          [CoolerArticleResource.detailShape().getFetchKey(params)]: {
            results: [5, 6, 7],
          },
        },
      };
      const { result } = renderHook(() =>
        useDenormalized(CoolerArticleResource.detailShape(), params, state),
      );
      expect(result.error).toBeDefined();
    });
    describe('nested resources', () => {
      const nestedArticle = NestedArticleResource.fromJS({
        ...params,
        user: 23,
      });
      const user = UserResource.fromJS({ id: 23, username: 'anne' });

      const state = {
        ...initialState,
        entities: {
          [NestedArticleResource.key]: {
            [`${nestedArticle.pk()}`]: nestedArticle,
          },
          [UserResource.key]: { [`${user.pk()}`]: user },
        },
      };
      state.entityMeta = createEntityMeta(state.entities);
      const {
        result: {
          current: { data, expiryStatus, expiresAt },
        },
      } = renderHook(() =>
        useDenormalized(NestedArticleResource.detailShape(), params, state),
      );
      it('expiryStatus should be Valid', () => {
        expect(expiryStatus).toBe(ExpiryStatus.Valid);
      });

      it('should provide inferred results', () => {
        expect(data).toMatchInlineSnapshot(`
          NestedArticleResource {
            "author": null,
            "content": "head",
            "id": 5,
            "tags": [],
            "title": "bob",
            "user": UserResource {
              "email": "",
              "id": 23,
              "isAdmin": false,
              "username": "anne",
            },
          }
        `);
      });
    });
  });

  describe('List', () => {
    const params = { things: 5 };
    const articles = [
      CoolerArticleResource.fromJS({ id: 5 }),
      CoolerArticleResource.fromJS({ id: 6 }),
      CoolerArticleResource.fromJS({ id: 34, title: 'five' }),
    ];
    describe('state is empty', () => {
      const state = initialState;
      const {
        result: {
          current: { data, expiryStatus, expiresAt },
        },
      } = renderHook(() =>
        useDenormalized(PaginatedArticleResource.listShape(), {}, state),
      );

      it('expiryStatus should be InvalidIfStale', () => {
        expect(expiryStatus).toBe(ExpiryStatus.InvalidIfStale);
      });

      it('should provide inferred results with undefined for entity', () => {
        expect(data).toMatchInlineSnapshot(`
          {
            "nextPage": "",
            "prevPage": "",
            "results": undefined,
          }
        `);
      });
    });
    describe('state is partial', () => {
      const { entities } = normalize(
        articles,
        CoolerArticleResource.listShape().schema,
      );
      const state = initialState;
      const {
        result: {
          current: { data, expiryStatus, expiresAt },
        },
      } = renderHook(() =>
        useDenormalized(CoolerArticleResource.listShape(), {}, state),
      );

      it('expiryStatus should be InvalidIfStale', () => {
        expect(expiryStatus).toBe(ExpiryStatus.InvalidIfStale);
      });

      it('should provide inferred results with undefined for entity', () => {
        expect(data).toMatchInlineSnapshot(`undefined`);
      });
    });
    describe('state exists', () => {
      const { entities, result: resultState } = normalize(
        articles,
        CoolerArticleResource.listShape().schema,
      );
      const state = {
        ...initialState,
        entities,
        entityMeta: createEntityMeta(entities),
        results: {
          [CoolerArticleResource.listShape().getFetchKey(params)]: resultState,
        },
      };
      const {
        result: {
          current: { data, expiryStatus, expiresAt },
        },
      } = renderHook(() =>
        useDenormalized(CoolerArticleResource.listShape(), params, state),
      );

      it('found should be true', () => {
        expect(expiryStatus).toBe(ExpiryStatus.Valid);
      });

      it('should provide inferred results', () => {
        expect(data).toStrictEqual(articles);
      });
    });
    describe('missing some ids in entities table', () => {
      const { entities, result: resultState } = normalize(
        articles,
        CoolerArticleResource.listShape().schema,
      );
      delete entities[CoolerArticleResource.key]['5'];
      const state = {
        ...initialState,
        entities,
        entityMeta: createEntityMeta(entities),
        results: {
          [CoolerArticleResource.listShape().getFetchKey(params)]: resultState,
        },
      };
      const {
        result: {
          current: { data, expiryStatus, expiresAt },
        },
      } = renderHook(() =>
        useDenormalized(CoolerArticleResource.listShape(), params, state),
      );

      const expectedArticles = articles.slice(1);

      it('expiryStatus should be Valid', () => {
        expect(expiryStatus).toBe(ExpiryStatus.Valid);
      });

      it('should simply ignore missing entities', () => {
        expect(data).toEqual(expectedArticles);
      });
    });
    describe('paginated results + missing some ids in entities table', () => {
      const { entities, result: resultState } = normalize(
        { results: articles },
        PaginatedArticleResource.listShape().schema,
      );
      delete entities[PaginatedArticleResource.key]['5'];
      const state = {
        ...initialState,
        entities,
        entityMeta: createEntityMeta(entities),
        results: {
          [PaginatedArticleResource.listShape().getFetchKey(params)]:
            resultState,
        },
      };
      const {
        result: {
          current: { data, expiryStatus, expiresAt },
        },
      } = renderHook(() =>
        useDenormalized(PaginatedArticleResource.listShape(), params, state),
      );

      it('expiryStatus should be Valid', () => {
        expect(expiryStatus).toBe(ExpiryStatus.Valid);
      });

      it('should match normalized articles', () => {
        const expectedArticles = articles.slice(1);
        expect(data.results).toEqual(expectedArticles);
      });
    });
    describe('paginated results', () => {
      const { entities, result: resultState } = normalize(
        { results: articles },
        PaginatedArticleResource.listShape().schema,
      );
      const state = {
        ...initialState,
        entities,
        entityMeta: createEntityMeta(entities),
        results: {
          [PaginatedArticleResource.listShape().getFetchKey(params)]:
            resultState,
        },
      };
      let result: any;
      const denormalizeCache = {
        entities: {},
        results: {},
      };

      beforeEach(() => {
        const v = renderHook(() => {
          const [usedState, setState] = useState(state);
          return {
            ret: useDenormalized(
              PaginatedArticleResource.listShape(),
              params,
              usedState,
              denormalizeCache,
            ),
            setState,
          };
        });
        result = v.result;
      });

      it('expiryStatus should be Valid', () => {
        expect(result.current.ret.expiryStatus).toBe(ExpiryStatus.Valid);
      });

      it('should match normalized articles', () => {
        expect(result.current.ret.data.results).toEqual(articles);
      });

      it('should stay referentially equal with external entity changes', () => {
        const prevValue = result.current.ret.data;
        act(() =>
          result.current.setState((state: any) => ({
            ...state,
            entities: { ...state.entities, whatever: {} },
          })),
        );
        expect(result.current.ret.data).toBe(prevValue);
        expect(result.current.ret.data.results).toBe(prevValue.results);

        act(() =>
          result.current.setState((state: any) => {
            const ret = {
              ...state,
              entities: {
                ...state.entities,
                [PaginatedArticleResource.key]: {
                  1430: 'fake2',
                  ...state.entities[PaginatedArticleResource.key],
                  100000: 'fake',
                },
              },
            };
            return { ...ret, entityMeta: createEntityMeta(state.entities) };
          }),
        );
        expect(result.current.ret.data).toBe(prevValue);
        expect(result.current.ret.data.results).toBe(prevValue.results);
      });

      it('should referentially change when an entity changes', () => {
        const prevValue = result.current.ret.data;
        act(() =>
          result.current.setState((state: any) => ({
            ...state,
            entities: {
              ...state.entities,
              [PaginatedArticleResource.key]: {
                ...state.entities[PaginatedArticleResource.key],
                '5': CoolerArticleResource.fromJS({ id: 5, title: 'five' }),
              },
            },
          })),
        );
        expect(result.current.ret.data).not.toBe(prevValue);
      });

      it('should referentially change when the result extends', () => {
        const prevValue = result.current.ret.data;
        act(() =>
          result.current.setState((state: any) => ({
            ...state,
            results: {
              ...state.results,
              [PaginatedArticleResource.listShape().getFetchKey(params)]: {
                results: [...(resultState.results ?? []), '5'],
              },
            },
          })),
        );
        expect(result.current.ret.data).not.toBe(prevValue);
        expect(result.current.ret.data).toMatchSnapshot();
      });
    });

    describe('paginated results still loading', () => {
      const { entities, result: resultState } = normalize(
        { results: articles },
        PaginatedArticleResource.listShape().schema,
      );
      const state = {
        ...initialState,
        entities,
        entityMeta: createEntityMeta(entities),
      };
      const {
        result: {
          current: { data, expiryStatus, expiresAt },
        },
      } = renderHook(() =>
        useDenormalized(PaginatedArticleResource.listShape(), params, state),
      );

      it('expiryStatus should be InvalidIfStale', () => {
        expect(expiryStatus).toBe(ExpiryStatus.InvalidIfStale);
      });

      it('value should be inferred for pagination primitives', () => {
        expect(data).toMatchInlineSnapshot(`
          {
            "nextPage": "",
            "prevPage": "",
            "results": undefined,
          }
        `);
      });
    });

    it('should not infer with schemas that have no entities', () => {
      const userId = '5';
      const { result } = renderHook(() => {
        return useDenormalized(photoShape, { userId }, initialState as any);
      });
      expect(result.current).toStrictEqual({
        data: null,
        expiresAt: 0,
        expiryStatus: ExpiryStatus.InvalidIfStale,
      });
    });

    it('should return results as-is for schemas with no entities', () => {
      const userId = '5';
      const results = new ArrayBuffer(10);
      const state = {
        ...initialState,
        results: {
          ...initialState.results,
          [photoShape.getFetchKey({ userId })]: results,
        },
      };
      const { result } = renderHook(() => {
        return useDenormalized(photoShape, { userId }, state);
      });
      expect(result.current).toStrictEqual({
        data: results,
        expiresAt: 0,
        expiryStatus: ExpiryStatus.Valid,
      });
    });

    it('should throw with invalid schemas', () => {
      const shape = PaginatedArticleResource.listShape();
      shape.schema = { happy: { go: { lucky: 5 } } } as any;
      const { result } = renderHook(() =>
        useDenormalized(shape, params, {} as any),
      );
      expect(result.error).toBeDefined();
    });
  });
});
