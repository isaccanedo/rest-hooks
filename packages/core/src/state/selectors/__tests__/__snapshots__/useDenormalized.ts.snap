// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useDenormalized() List paginated results should referentially change when the result extends 1`] = `
{
  "results": [
    PaginatedArticleResource {
      "author": null,
      "content": "",
      "id": 5,
      "tags": [],
      "title": "",
    },
    PaginatedArticleResource {
      "author": null,
      "content": "",
      "id": 6,
      "tags": [],
      "title": "",
    },
    PaginatedArticleResource {
      "author": null,
      "content": "",
      "id": 34,
      "tags": [],
      "title": "five",
    },
    PaginatedArticleResource {
      "author": null,
      "content": "",
      "id": 5,
      "tags": [],
      "title": "",
    },
  ],
}
`;
